### ISSUE TITLE
Host Github by itself as open source project

### ISSUE THREAD

**Issue by joelhandwell** [dear-github/dear-github#304]
Microsoft [open sourced Xamarin](https://www.theregister.co.uk/2016/03/31/xamarin_tools_code_free_and_open_source/) after acquisition and it was good for community. Github was not open source for its own reason and it could be same reason why Xamarin was not. Now that Github acquired by Microsoft, would this event be an opportunity to re-consider the reason?

**Comment by aethelwulffe** [comment]
And add a tag to all comments by MS employees.
MS lost me a long time ago (the Hotmail acquisition and the resulting "eh" password scandal in the 90's).  Since then, Windows has been this thing that I have to spend hours stripping the junk out of, adding service packs to, holding my mouth right to install licenses, hoping that nothing bad will happen to configurations that do not need or want any "Creator Apps", and dreading and fearing that my motherboard will die and prevent me from accessing my invested licensed software on a windows OS hard-drive that I can't just stick in another machine and use.

Dudes.  There is a LOT of bad blood between the larger Geek community and the "Gates'ed Community".  LOTS.  I am a (mostly) Windows user because I have to be to competently support the mental health and medical community.  If not for that, I would be in the overwhelming majority of Linux-only people.  I get snickered at for not showing pure hate and disdain for MS, or M$ as it is known (seriously, who does not know what "M$" stands for?)

-So.  Totally.  Don't.  Screw.  This.  Up.

People will BAIL and head for the shitty UI of GitLAB in a heartbeat.  I have fought our own project's move there.  Making the actual Github application software open-source would:
1.  Be a dream solution for a lot of folks.
2.  Be a dream solution that would actually KEEP FOLKS HERE, because they could have confidence that they can always move into their own box later (why bother at that point?)
3.  Be the most anticipated and popular repo on the planet.  WHO WOULDN'T want to have a fork?

**Comment by duttaditya18** [comment]
Maybe not. GitHub's main idea is Social Coding and giving users a single platform for contributing to several Open Source projects. Having everyone host their own version of GitHub kind of defeats this purpose.

So, I don't think this will happen.

**Comment by dminkovsky** [comment]
Xamarin was a framework, GitHub is a social network.

**Comment by k3170makan** [comment]
We should make them understand that if developers developer developers are so damn important then they should let us see the code they forcing us to run and our users.

**Comment by patrickcurl** [comment]
@duttaditya18 It could still be social in nature, make MS Github be an aggregator, with all other GH's being either federated (self-hosted but public) or centrally-hosted(on gh servers), you could search for a project, star them, etc...and hit either central OR federated, federated servers could do all things the main one can, plus more (w/ custom plugins/addons/etc)...

**Comment by nixpulvis** [comment]
Make something called GitPub, and have GitHub be the first de-facto implementation. Boom, problem solved. There will always be value in the features (UI/UX) of the client, but the idea a single domain owns the entire social element isn't going to stand the test of time.

**Comment by santiagobasulto** [comment]
I don't want to get into the discussion of "should" or "should not" (or will/won't, would/wound't) open source the code.

I'm just rooting for this because it'd be GREAT from an educational perspective. As a teacher, I love to find real life applications of the things I teach and point my students to them.

So, a solid +1 to this one.

**Comment by sam0x17** [comment]
What needs to happen is someone needs to take all the functionality github offers and implement it as addon features for git that all run from a terminal.

**Comment by sam0x17** [comment]
> Open-sourcing GitHub would take away business from GitHub Enterprise, which is probably why Microsoft purchased GitHub in the first place.

If you think Microsoft purchased GitHub expecting to make money directly from GitHub, I think you are missing the point. They didn't buy a profitable business, they bought what they at least thought was the hearts and minds of developers everywhere, and more importantly an extremely good platform for advertising and promoting Azure.

**Comment by joelhandwell** [comment]
I'm totally not interested in running my own github instance in other server just like people who signed the open letter dear-github to plead Github to add +1 button in issue which took really long to be recognized as an issue to be implemented. I'm sure if it was open source, somebody implemented the feature and made a pull request.

My intention is to make github.com "Client Side Code" better by letting OSS community join into the development. So open sourcing backend and architecture is not my interest and possibly not interest of MS too. I'm not sure if Github is made of micro services, but if it is, certain client side code could be open sourced first and only that could generate great contribution to improve it.

I'm heavy user of Gitlab and disappointed because it is no longer fully open source and some enterprise features would not be welcomed to be implemented by OSS community because it conflicts with Gitlab company revenue. 

In my ideal world, MS splits Github source into client and backend and open source "Github client + test backend" for community to test if client code works while they make pull request and keep production backend closed. 

**Update:**
[TheHydroImpulse in HackerNews gave insights](https://news.ycombinator.com/item?id=17448076) on how it could be difficult to split client and backend of Github for just open source sake:

> Github has been a well known Rails monolith. Absolutely they have services that power all sorts of stuff (they have a few blog posts on it) but it's just impractical to even start a discussion on splitting it up for the reasons you initially argued for. I've been through a number of large rewrites/reworks that took monoliths much like Github (with many many many services behind it) and split them up into modular pieces and it's an insane amount of work that can take years. You simply need very good reasons (including business reasons) to do that. Moreover, companies at these sizes just have a LOT of code all over the place. Tooling, infra, supporting services, etc... 

> Not to mention it's just not useful to have external contributors for a business product like Github. Doing code reviews, addressing bugs that were introduced, spending time discussing things with contributors takes an incredible amount of time.Basically if the reason you want Github open sourced (and reworked into some weird architecture you described) is so that people can contribute to fix things and add features.... Github could/will just hire more devs to work on that.

**Comment by notklaatu** [comment]
Microsoft has proclaimed very loudly that they are now a company that embraces open source. Making GitHub an open source project would be the _action_ to their _words_.

**Comment by joelhandwell** [comment]
@satyanadella @jeffmcaffer @natfriedman @defunkt @bkeepers any thoughts?

**Comment by sam0x17** [comment]
@joelhandwell github is a rails app

**Comment by joelhandwell** [comment]
@sam0x17 Yes. [I came to know that](https://github.com/dear-github/dear-github/issues/304#issuecomment-402010067) and it's monolith... So it will be difficult to split the code.

**Comment by joelhandwell** [comment]
Thinking about the fact that Github have Github.com (production service) and Github Enterprise (self hosted), could it be like this?

- Github.com = Github Core + Production Services and Infra
- Github Enterprise = Github Core + Services need for Self Hosting

Maybe it's not worthy to proceed based on assumption but if there is something like "Github Core" which is shared codebase between prod and self-hosted, open sourcing the core can be an option?

**Comment by k3170makan** [comment]
Why is there even a split between making money and showing people the code you are forcing them to run? Is it not an ethical issue on its own? And is it not unethical to cast ethical problems out of ethical domains fundamentally?

**Comment by aethelwulffe** [comment]
@k3170makan  Hi Keith, Art here.

I think it is a bit of a stretch to say "Forcing you to run".  I am pretty sure that you can create your own git server that just about any terminal or git-gui can interact with.  You can also install and use a huge variety of forum, issue tracking software etc... (Bugzilla does exist you know...wave to Dylan!) that major enterprises use and fork until their heart is content.  What we have here at GitHUB is really a social media platform that hands you all that stuff in a ready-to-use package.

-I have not noticed a lot of adverts popping up on GitHUB, and despite the fact that it practically owns my email client on a volume basis (my own server + good old Thunderbird), I have never received any traffic on my mail server that could be identified as spam associated with this account.  Moreover, as an auth supplier, it has probably protected me from junk other places.  I am not on Facebook or any of that other junk (OK, Linkedin, which IS a source of junk), and I have always been quite happy to pay about $50 a month for a handful of seats for a private repo _that I don't actually use much._  That is right.  I don't really need a private repo, but I find it is a handy way to pay for services I use heavily.

So, I am on public repos, but I pay for this service, Part of my project organization uses GitLAB.  I find it a poor substitute.  Sourceforge showed itself to be pure evil (and I dearly loved Sourceforge).  To have all this stuff just exist out there, it needs to be paid for.  I thought (assumed) that people understood that they needed to support this company, and were giving them a decent level of subscriptions.  I figured that they needed to keep the source to themselves 1.  Because one of the big boys would be able to use it to finally figure out how this should be done and then media-blitz their own project to destroy, in preparation for a bait and switch.
2.  The "software" was actually a pile of things (not just this referential issue tracker) that would be impractical for anyone not at that level to implement anyway.

Besides which..This is a rails app.  It is amazing that it runs at all.  Seeing the code would just make me hurl.  😈

**Comment by joelhandwell** [comment]
After Github source hosted on Github, the list of issues in the internal closed issue tracker can be merged with issues in [isaacs/github](https://github.com/isaacs/github)

**Comment by DamieFC** [comment]
I feel that Github should be made open source because it promotes open source and sharing code. With it not being open source, you're promoting something that you're not. It's wrong.

**Comment by xianchaoyu** [comment]
mark
